Run pulp content repository on the repo host.

nuclearsandwich opened the pull request:
Run the [pulp](https://pulpproject.org) content repository software in containers on the repo host.
This branch uses docker for the contained services. I struggled for a while to try and get a podman based approach in order to leverage rootless and daemonless containers but ran into constant challenges with the effective user and environment which where not yielding to reasonable seeming solutions and I was running out of motivation to figure out what the heck was going on. The abandoned pulp branch is https://github.com/ros-infrastructure/cookbook-ros-buildfarm/tree/nuclearsandwich/podman-pulp-wip

In an earlier iteration of this project the original repository for systemd-docker was not working and unmaintained. I found a maintained fork and have been successfully using the binary provided there.

nuclearsandwich committed:
Reapply "WIP on pulp support for repo recipe."

This reverts commit 9d9c03c33bd6ce4e72a00535a6d3d94f01d2641d to resume
work on Pulp.

nuclearsandwich committed:
Install Podman for running contained pulp services.

The initial plan was to use systemd-docker but it appears to have
fallen out of maintenance and it's unclear if there is a successor
project with as much simplicity.

Podman's own systemd integration also proved to have caveats, largely
around the legibility of its generated systemd units which have names
like `container-OBNOXIOUSLYLONGCONTAINERID` instead of generating units
based on the container name. I ended up using the systemd unit @cottsay
and I wrote for systemd-docker as a starting point and folding in
Podman's techniques for managing container supervision. The result is
legible although not as neat as systemd-docker, now it just has to work,
which is what proved difficult for systemd-docker.

One difference between podman as we're using it and docker is that the
container is pre-created and in a ready state before being started in
the service unit whereas in the systemd-docker case the container is
created and removed by the unit when started and stopped.

nuclearsandwich committed:
WIP: Add containers and services for pulp.

I'm having quite the challenge balancing the way chef is invoked within
test kitchen and environmment leaking, along with getting the
permissions together for rootless container invocation.

I'm also not confident that I've got uid 1200 in the container mapped to
1200 outside it I think I need an extra setuid/setgid table entry

nuclearsandwich committed:
Pod progress is made but not complete.

nuclearsandwich committed:
Start dockering the podman.

I hope never to return here.

nuclearsandwich committed:
Heck yeah! A systemd-docker service is running!

nuclearsandwich committed:
Use Docker with systemd-docker for pulp containers.

nuclearsandwich committed:
Remove policy.json file used by podman and buildah.

nuclearsandwich commented:
@cottsay if the setup here looks reasonably complete I'll go ahead and deploy it to the test build farm. I'm aware there are still some TODOs about getting the gpg stuff set up for pulp and lingering comments (of mine) for creating pulp workers and creating the rpm repos. I think the workers comment is probably stale given that pulp workers are created just above it but I haven't returned to https://github.com/ros-infrastructure/buildfarm_deployment/pull/228/ to port over configuring and creating the RPM repositories.

nuclearsandwich commented:
The CI failures are due to the action running test kitchen inside docker. @j-rivero do you know whether dokken allows you to bind-mount the host docker socket so that sibling containers can be spawned?

nuclearsandwich committed:
Try allowing sibling docker containers to be created.

cottsay commented on files/rpm_repo_init.py:
```diff
@@ -0,0 +1,220 @@
+from urllib3.exceptions import MaxRetryError
+
+
+def _get_recursive_dependencies(packages, target_names):
+    to_remove = dict()
+    new_names = set(target_names)
+
+    while new_names:
+        target_names = new_names
+        new_names = set()
+
+        for pkg in packages:
+            if [r for r in pkg.requires if r[0] in target_names]:
+                to_remove[pkg.pulp_href] = pkg
+                new_names.add(pkg.name)
+
+    return to_remove
+
+
```
```suggestion
```

Looks like this can actually be dropped. I don't see it referenced.

cottsay commented on kitchen.dokken.yml:
```diff
@@ -5,6 +5,7 @@ driver:
   use_sudo: false
   privileged: true
   env: [CHEF_LICENSE=accept]
+  volumes: [ '/var/lib/docker', '/var/run/docker.sock' ]
```
```suggestion
  volumes: ['/var/lib/docker', '/var/run/docker.sock']
```

cottsay commented on templates/pulp/pulp.service.erb:
```diff
@@ -0,0 +1,17 @@
+Require=docker.service <%= @required_units.join(' ') %>
+
+[Service]
+ExecStart=/usr/local/bin/systemd-docker --cgroups name=systemd run --rm <%= @docker_create_args %> --name %p%i  pulp_image <%= @docker_cmd %>
```
```suggestion
ExecStart=/usr/local/bin/systemd-docker --cgroups name=systemd run --rm <%= @docker_create_args %> --name %p%i pulp_image <%= @docker_cmd %>
```

cottsay commented on recipes/repo.rb:
```diff
@@ -265,6 +265,162 @@
+cookbook_file "#{pulp_data_directory}/initialize.py" do
+  owner 'pulp'
+  source 'rpm_repo_init.py'
+  mode '0600'
```
If we're going to execute the script explicitly with the interpreter, we should drop the shebang from `rpm_repo-init.py`.

nuclearsandwich replied:
Do you have a preference for executable vs otherwise? I don't.

cottsay replied:
In the spirit of deterministic behavior, I'd guess my preference would be to invoke it using the interpreter directly, and not to rely on the shebang. But I don't really care very much either, just looking for consistency.

cottsay commented on recipes/repo.rb:
```diff
@@ -265,6 +265,162 @@
+
+pulp_admin_password = data_bag_item('ros_buildfarm_pulp_admin_password', node.chef_environment)['password']
+execute 'set_pulp_admin_password' do
+  command "docker run --user 1200 --rm -v /var/run/postgresql:/var/run/postgresql pulp_image pulpcore-manager reset-admin-password -p '#{pulp_admin_password}'"
```
Note that this step was missing shell escaping for the password in the original PR. The randomly generated password I used on the test farm broken the step.

cottsay commented on recipes/repo.rb:
```diff
@@ -265,6 +265,162 @@
+    description: "Pulp API Endpoint",
+    after_units: %w(postgresql.service redis-server),
+    required_units: %w(postgresql.service redis-server),
+    docker_create_args: %(-u 1200 -v #{pulp_data_directory}:/var/repos/.pulp -v /var/run/postgresql:/var/run/postgresql -v /var/run/redis:/var/run/redis -p 24817:24817),
```
```suggestion
    docker_create_args: %(-u 1200:1200 -v #{pulp_data_directory}:/var/repos/.pulp -v /var/run/postgresql:/var/run/postgresql -v /var/run/redis:/var/run/redis -p 24817:24817),
```

nuclearsandwich replied:
Right. I had dropped the : separator because in the podman work `1200:1200` indicates run the process with uid 1200 inside the container mapped to uid 1200 on the host whereas in docker it indicates the usual uid:gid.

cottsay reviewed (commented):
A few comments, but this looks like a good start, and the bulk of what is needed to move forward.

nuclearsandwich committed:
Use an attribute to disable pulp services.

The pulp services are not currently runnable via kitchen-dokken and I
haven't been able to figure out a reliable way to fix that.

nuclearsandwich committed:
Remove mounts from kitchen.dokken.yml.

nuclearsandwich committed:
Disable pulp services on repo host.

nuclearsandwich committed:
Remove extra space.

Co-authored-by: Scott K Logan <logans@cottsay.net>

nuclearsandwich committed:
Fix missing gid map in docker -u argument.

nuclearsandwich committed:
Remove unused function.

Co-authored-by: Scott K Logan <logans@cottsay.net>

cottsay commented on recipes/repo.rb:
```diff
@@ -265,6 +265,163 @@
+    command "docker run --user 1200 --rm -v #{pulp_data_directory}:/var/repos/.pulp pulp_image pulpcore-manager collectstatic --noinput"
+  end
+
+  pulp_admin_password = data_bag_item('ros_buildfarm_pulp_admin_password', node.chef_environment)['password']
```
It would be good to get this credential populating on the Jenkins master. Once we have that and the pulp client pip packages installed on the master and all of the workers, the remaining changes will all be on the repo host.

nuclearsandwich replied:
Updated in 
2987f5a

nuclearsandwich committed:
Fix problems with pulp services.

The field is Requires= not Require= and the redis-server unit needs the
.service extension.

nuclearsandwich committed:
Run pulp commands with both uid and gid of pulp.

nuclearsandwich committed:
Fix missing - on -c flag.

nuclearsandwich committed:
Initialize RPM repositories.

nuclearsandwich committed:
Pin to the same version of pulp that we're running on the server.

nuclearsandwich committed:
Remove #! from file.

The expectation is that it is run as a script rather than an executable.

nuclearsandwich committed:
Use jenkins password credential for pulp password.

Jenkins will need access to the pulp password for some jobs so rather
than a separate data bag, use a password credential with a specific ID.

cottsay reviewed (approved):
I think this is a good/stable milestone.

It wouldn't be a bad idea to merge this as-is, deploy it to the staging farm, and iterate from there.

cottsay commented on recipes/agent.rb:
```diff
@@ -100,7 +100,7 @@
 # Client utilities for pulp content manager
 package "python3-pip"
-%w[pulp-rpm-client==3.8.0 pulpcore-client==3.8.1].each do |pulppkg|
+%w[pulp-rpm-client==3.4.1 pulpcore-client==3.4.1].each do |pulppkg|
```
I did some testing, and it looks like the issues we were having with the newer Pulp versions were caused by a mismatch, and not necessarily our use of the API. Specifically, things seem to go awry if the client code is newer than the API server.

I found that `pulpcore==3.9.1` and `pulp-rpm==3.9.0` (and their corresponding `-client` packages) seemed to work well in concert.

It would probably be worth upgrading everything prior to merging this and capturing the new AMI for the workers.

nuclearsandwich replied:
By upgrading everything do you mean updating the pulp version both on the agents and in the pulp image?

cottsay replied:
Correct

nuclearsandwich committed:
Bump pulp client versions.
